src/tqec/circuit/detector/detector_util.py

```diff
@@ -0,0 +1,406 @@
+        self.seen_measurement_at_tail = False
+
+    def to_fragment(self) -> Fragment:
+        return Fragment(
+            circuit=self.cur_fragment.copy(),
+            end_stabilizer_sources=list(self.end_stabilizer_sources),
+            begin_stabilizer_sources=list(self.begin_stabilizer_sources),
+            sources_for_next_fragment=list(self.sources_for_next_fragment),
+        )
+
+
+def split_stim_circuit_into_fragments(
+    circuit: stim.Circuit,
+) -> list[Fragment | FragmentLoop]:
+    """Split the circuit into fragments."""
+    fragments = []
+    state = SplitState()
+    for moment in iter_stim_circuit_by_moments(circuit):
+        if isinstance(moment, stim.CircuitRepeatBlock):
+            if state.cur_fragment:
+                fragments.append(state.to_fragment())
+                state.clear()
+            body_fragments = split_stim_circuit_into_fragments(moment.body_copy())
+            fragments.append(
+                FragmentLoop(fragments=body_fragments, repetitions=moment.repeat_count)
+            )
+        elif has_measurement(moment, check_are_all_measurements=True):
+            state.cur_fragment += moment
+            state.begin_stabilizer_sources.extend(
+                collapse_pauli_strings_at_moment(moment, is_reset=False)
+            )
+            state.seen_measurement_at_tail = True
+            state.sources_for_next_fragment.extend(
+                collapse_pauli_strings_at_moment(moment, is_reset=True)
+                if has_reset(moment, check_are_all_resets=False)
+                else []
+            )
+        elif has_reset(moment, check_are_all_resets=True):
+            pauli_strings = collapse_pauli_strings_at_moment(moment, is_reset=True)
+            if state.seen_measurement_at_tail:
+                state.cur_fragment += moment
+                state.sources_for_next_fragment.extend(pauli_strings)
+            elif state.seen_reset_at_head:
+                state.cur_fragment += moment
+                state.end_stabilizer_sources.extend(pauli_strings)
+            else:
+                if state.cur_fragment:
+                    fragments.append(state.to_fragment())
+                    state.clear()
+                state.cur_fragment += moment
+                state.end_stabilizer_sources.extend(pauli_strings)
+                state.seen_reset_at_head = True
+
+        elif state.seen_measurement_at_tail:
+            fragments.append(state.to_fragment())
+            state.clear()
+            state.cur_fragment += moment
+        else:
+            state.cur_fragment += moment
+    if state.cur_fragment:
+        fragments.append(state.to_fragment())
+    return fragments
+
+
+def pauli_string_mean_coords(
+    pauli_string: PauliString, qubit_coords_map: dict[int, list[float]]
+) -> tuple[float, ...]:
+    all_coords_items = [qubit_coords_map[i] for i in pauli_string.qubits.keys()]
+    return tuple(np.mean(np.asarray(all_coords_items), axis=0)) + (0.0,)
```

Comment:
What is the meaning of the last `+ (0.0,)`? Is it to add the time dimension?

Reply:
Yes it is the time dimension.